### PR TITLE
SPLAT-1129: adding failure domain awareness for vSphere

### PR DESCRIPTION
**Pull request by rvanderp3**
Associated PRs:
- [x] https://github.com/openshift/cluster-api-actuator-pkg/pull/288
- [x] https://github.com/openshift/api/pull/1539
- [x] https://github.com/openshift/client-go/pull/251
- [x] https://github.com/openshift/cluster-config-operator/pull/342
- [x] https://github.com/openshift/installer/pull/7418

Depends on PRs:
- [x] https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/250

### CODE REVIEW

**Comment by rvanderp3** [comment]
converting to draft to prevent unnecessary CI runs.

File: pkg/controllers/controlplanemachineset/suite_test.go [hunk 1]
```diff
@@ -86,6 +88,10 @@ var _ = BeforeSuite(func() {
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	infrastructure := configv1builder.Infrastructure().AsAWS("test", "eu-west-2").WithName(util.InfrastructureName).Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/suite_test.go:93]: Idiom nit
```suggestion
	Expect(k8sClient.Create(ctx, infrastructure)).To(Succeed())
```

File: pkg/controllers/controlplanemachinesetgenerator/suite_test.go [hunk 2]
```diff
@@ -84,6 +84,10 @@ var _ = BeforeSuite(func() {
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	/*infrastructure := configv1builder.Infrastructure().AsAWS("cluster", "us-east-1").WithName("cluster").Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())*/
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/suite_test.go:89]: Not needed?

File: pkg/webhooks/controlplanemachineset/suite_test.go [hunk 3]
```diff
@@ -77,13 +80,18 @@ var _ = BeforeSuite(func() {
...
 
 	//+kubebuilder:scaffold:scheme
 
 	k8sClient, err = client.New(cfg, client.Options{Scheme: testScheme})
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	infrastructure := configv1builder.Infrastructure().AsAWS("cluster", "us-east-1").WithName("cluster").Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/webhooks/controlplanemachineset/suite_test.go:93]: Idiom nit
```suggestion
	Expect(k8sClient.Create(ctx, infrastructure)).To(Succeed())
```

File: pkg/webhooks/controlplanemachineset/webhooks.go [hunk 4]
```diff
@@ -364,6 +378,12 @@ func validateOpenShiftOpenStackProviderConfig(parentPath *field.Path, providerCo
 	return []error{}
 }
 
+// validateOpenShiftVSphereProviderConfig runs VSphere specific checks on the provider config on the ControlPlaneMachineSet.
+// This ensure that the ControlPlaneMachineSet can safely replace VSphere control plane machines.
+func validateOpenShiftVSphereProviderConfig(parentPath *field.Path, providerConfig providerconfig.VSphereProviderConfig, infrastructure *configv1.Infrastructure) []error {
+	return []error{}
```
Comment (JoelSpeed) [inline pkg/webhooks/controlplanemachineset/webhooks.go:384]: Is there any validation planned? Anything we know must be configured in a vSphere ProviderSpec for it to work with control planes?

File: pkg/test/suite_test.go [hunk 5]
```diff
@@ -70,6 +72,10 @@ var _ = BeforeSuite(func() {
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	infrastructure := configv1builder.Infrastructure().AsAWS("test", "eu-west-2").WithName(util.InfrastructureName).Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/test/suite_test.go:77]: Idiom nit
```suggestion
	Expect(k8sClient.Create(ctx, infrastructure)).To(Succeed())
```

File: pkg/machineproviders/providers/suite_test.go [hunk 6]
```diff
@@ -81,6 +83,10 @@ var _ = BeforeSuite(func() {
 	Expect(err).NotTo(HaveOccurred())
 	Expect(k8sClient).NotTo(BeNil())
 
+	infrastructure := configv1builder.Infrastructure().AsAWS("test", "eu-west-2").WithName(util.InfrastructureName).Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/suite_test.go:88]: Idiom nit
```suggestion
	Expect(k8sClient.Create(ctx, infrastructure)).To(Succeed())
```

File: pkg/machineproviders/providers/openshift/machine/v1beta1/suite_test.go [hunk 7]
```diff
@@ -75,6 +77,10 @@ var _ = BeforeSuite(func() {
 	Expect(machinev1beta1.Install(testScheme)).To(Succeed())
 	Expect(configv1.Install(testScheme)).To(Succeed())
 
+	infrastructure := configv1builder.Infrastructure().AsAWS("test", "eu-west-2").WithName(util.InfrastructureName).Build()
+	err = k8sClient.Create(ctx, infrastructure)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/suite_test.go:82]: Idiom nit
```suggestion
	Expect(k8sClient.Create(ctx, infrastructure)).To(Succeed())
```

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+// It allows external code to extract and inject failure domain information,
+// as well as gathering the stored config.
+type VSphereProviderConfig struct {
+	providerConfig machinev1beta1.VSphereMachineProviderSpec
+	infrastructure *v1.Infrastructure
+}
+
+// InjectFailureDomain returns a new VSphereProviderConfig configured with the failure domain.
+//
+//nolint:nestif
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:40]: I'm not a fan of nolint in general, especially something like this one, a little refactor could probably mitigate this

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+func (v VSphereProviderConfig) InjectFailureDomain(fd machinev1.VSphereFailureDomain) VSphereProviderConfig {
+	newVSphereProviderConfig := v
+
+	for _, failureDomain := range v.infrastructure.Spec.PlatformSpec.VSphere.FailureDomains {
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:44]: Rather than a loop, what about a helper that gets the named failure domain out of the spec? And returns an error when not found?

Speaking of, that's a new thing, what happens if the failure domain that's named, isn't present in the infrastructure?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+			workspace.Datastore == topology.Datastore &&
+			workspace.Server == failureDomain.Server &&
+			workspace.ResourcePool == topology.ResourcePool {
+			return machinev1.VSphereFailureDomain{
+				Name: failureDomain.Name,
+			}
+		}
+	}
+
+	return machinev1.VSphereFailureDomain{}
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:114]: I would have thought we would consider this an error? 

I'm trying to work out what happens if someone were to create a cluster, add a new failure domain, then remove and old failure domain, have you tested this?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/failuredomain/failuredomain_test.go [hunk 11]
```diff
@@ -173,6 +173,49 @@ var _ = Describe("FailureDomains", func() {
...
+				failureDomains, err = NewFailureDomains(config)
+			})
+
+			It("should not error", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("should construct a list of failure domains", func() {
+				Expect(failureDomains).To(ConsistOf(
+					HaveField("String()", "VSphereFailureDomain{Zone:us-central1-a}"),
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/failuredomain/failuredomain_test.go:192]: Why does the string represent it as a Zone? Where did that come from? Is this not the failure domain name?

**Review by JoelSpeed — COMMENTED** [review]
Do we have an E2E test environment for multi zonal vSphere? Would be great to run the E2E suite for CPMS with vSphere if possible to build some confidence in it working as expected

CC @damdo @odvarkadaniel I'd appreciate if you have time to review this as well to see if you have any concerns about the inject/extract failure domains stuff in particular

**Comment by rvanderp3** [comment]
> Do we have an E2E test environment for multi zonal vSphere? Would be great to run the E2E suite for CPMS with vSphere if possible to build some confidence in it working as expected
> 
> CC @damdo @odvarkadaniel I'd appreciate if you have time to review this as well to see if you have any concerns about the inject/extract failure domains stuff in particular

yes, we do have a multi-zone vsphere environment and we have been performing the bulk of our testing there actually.   
 We've been able to confirm that failure domains are being honored.  I'll work on getting a run with the e2e suite.

File: pkg/webhooks/controlplanemachineset/webhooks.go (hunk 4 above)
Comment (rvanderp3) [reply pkg/webhooks/controlplanemachineset/webhooks.go:384]: yes, we are planning to implement validation

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 8 above)
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:40]: ack, you're right, we are working on the refactor

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 10 above)
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:138]: @JoelSpeed attached is a link to a spike on this question https://issues.redhat.com/browse/SPLAT-1185 .

TL;DR if a failure domain is removed from the infrastructure resource, the control plane machine set will set an Error condition stating that it encountered an unknown failure domain.  If the unknown failure domain is removed from the cpms instance, the node on the removed failure domain is replaced with a node in a valid failure domain.

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig_test.go [hunk 12]
```diff
@@ -556,6 +657,7 @@ var _ = Describe("Provider Config", func() {
 			comparePC     ProviderConfig
 			expectedEqual bool
 			expectedError error
+			bp            bool
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig_test.go:660]: Assuming this is for debugging?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+	return workspace
+}
+
+// InjectFailureDomain returns a new VSphereProviderConfig configured with the failure domain.
+func (v VSphereProviderConfig) InjectFailureDomain(fd machinev1.VSphereFailureDomain) (VSphereProviderConfig, error) {
+	newVSphereProviderConfig := v
+
+	failureDomain, err := newVSphereProviderConfig.getFailureDomainFromInfrastructure(fd)
+	if err != nil {
+		return newVSphereProviderConfig, errors.Wrap(err, "unknown failure domain")
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:95]:
```suggestion
		return newVSphereProviderConfig, fmt.Errorf("unknown failure domain: %w", err)
```

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+					NetworkName: topology.Networks[0],
+				},
+			},
+		}
+	}
+
+	if len(topology.Template) > 0 {
+		newVSphereProviderConfig.providerConfig.Template = topology.Template[strings.LastIndex(topology.Template, "/")+1:]
+	} else if len(v.infrastructure.Spec.PlatformSpec.VSphere.FailureDomains) > 1 {
+		newVSphereProviderConfig.providerConfig.Template = fmt.Sprintf("%s-rhcos-%s-%s", v.infrastructure.Status.InfrastructureName, failureDomain.Region, failureDomain.Zone)
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:114]: This seems like a little bit of magic to me, what's this? Can we add a comment to explain why there's this fallback for an unset template?

What happens if the template by this name doesn't exist? Wondering if it would be safer to error when the template is empty and use that as in instruction to users that they must populate the required field

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 10 above)
Comment (JoelSpeed) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:145]: And the code as it is today supports that?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	return newVSphereProviderConfig, nil
+}
+
+// TO-DO: should we return an error here? a failure domain could fail to be extracted for
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:120]: How we doing on this TODO?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere_test.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+	Context("ExtractFailureDomain", func() {
+		It("returns the configured failure domain", func() {
+			expected := machinev1resourcebuilder.VSphereFailureDomain().
+				WithZone(usCentral1a).
+				Build()
+
+			Expect(providerConfig.ExtractFailureDomain()).To(Equal(expected))
+		})
+
+		logger = testutils.NewTestLogger()
```
Comment (JoelSpeed) [inline pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere_test.go:59]: That's an odd place for this declaration? Are we using the test logger in this context?

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 14 above)
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:114]: If it's not defined, we revert to the logic that the installer follows https://github.com/openshift/installer/blob/9063acbd19767c5eb24e81b8f57a4b8717f02a2b/pkg/asset/machines/vsphere/machines.go#L80-L84 .  Sure, we'll add a comment.

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig.go [hunk 17]
```diff
@@ -28,6 +27,7 @@ import (
 	machinev1 "github.com/openshift/api/machine/v1"
 	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
 	"github.com/openshift/cluster-control-plane-machine-set-operator/pkg/machineproviders/providers/openshift/machine/v1beta1/failuredomain"
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig.go:30]: this was pulled in so errors could be wrapped, but we'll revert to fmt.errorf

File: pkg/controllers/controlplanemachinesetgenerator/vsphere.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+	emptyFailureDomain := machinev1.FailureDomains{}
+
+	// We want to make sure that if a failure domain is empty, it is the only one.
+	if len(VSphereFailureDomains) > 1 {
+		for _, fd := range VSphereFailureDomains {
+			if fd == emptyVSphereFailureDomain {
+				return emptyFailureDomain, fmt.Errorf("error building VSphere failure domains: %w", errMixedEmptyFailureDomains)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/vsphere.go:106]: yep, you're right.  thanks.

File: pkg/controllers/controlplanemachinesetgenerator/vsphere.go [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+	// We want to make sure that if a failure domain is empty, we don't create it.
+	if len(VSphereFailureDomains) == 1 && VSphereFailureDomains[0] == emptyVSphereFailureDomain {
+		VSphereFailureDomains = nil
+	}
+
+	return machinev1.FailureDomains{
+		VSphere:  VSphereFailureDomains,
+		Platform: configv1.VSpherePlatformType,
+	}, nil
```
[the remark this answers is not in the document]
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/vsphere.go:116]: you could conceivably have no failure domains, especially on an upgraded cluster that was born in 4.12 or prior.  The proposed API for [FailureDomains allows the slice of failure domains to remain optional.](https://github.com/openshift/api/pull/1539/commits/ecbb849a80a2d8b9e9aba37638a02fcc932b5b58#diff-f988b6561cddb07d3d89440736ea687f1d93bc918f71fafa3022614f1c34441bR257)

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 15 above)
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:120]: it is still WIP, we'll get to it soon

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig_test.go [hunk 20]
```diff
@@ -256,6 +301,34 @@ var _ = Describe("Provider Config", func() {
...
+					platformType: configv1.VSpherePlatformType,
+					vsphere: VSphereProviderConfig{
+						providerConfig: *machinev1beta1resourcebuilder.VSphereProviderSpec().WithZone("us-central1-a").Build(),
+						infrastructure: vsphereInfrastructureWithFailureDomains,
+					},
+				},
+				failureDomain: failuredomain.NewVSphereFailureDomain(
+					machinev1.VSphereFailureDomain(machinev1resourcebuilder.VSphereFailureDomain().WithZone("us-central1-a")),
+				),
+				failureDomainFunc: getAppliedVSphereFailureDomain,
```
[the remark this answers is not in the document]
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/providerconfig_test.go:315]: yep, thanks

File: pkg/controllers/controlplanemachinesetgenerator/vsphere.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/vsphere.go:104]: makes sense, thanks for the feedback Joel.  we'll be sure to follow that guidance.

**Comment by rvanderp3** [comment]
/hold

placing hold pending tests and reviews

**Comment by rvanderp3** [comment]
make unit

**Comment by rvanderp3** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.15-e2e-vsphere-ovn-techpreview-serial

**Comment by rvanderp3** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.15-e2e-vsphere-ovn-serial

**Comment by rvanderp3** [comment]
vSphere jobs are being impacted by etcd performance issues.  The infrastructure team is aware and is invstigating.

File: Makefile [hunk 21]
```diff
@@ -83,7 +83,7 @@ vendor: ## Ensure the vendor directory is up to date.
 
 .PHONY: lint
 lint: ## Run golangci-lint over the codebase.
-	$(call ensure-home, ${GOLANGCI_LINT} run ./... --timeout 5m)
+	$(call ensure-home, GOGC=50 ${GOLANGCI_LINT} run ./... --timeout 5m --new-from-rev=${PULL_BASE_SHA} -v)
```
Comment (JoelSpeed) [inline Makefile:86]: Does the linter still work locally with these changes?

File: cmd/control-plane-machine-set-operator/main.go [hunk 22]
```diff
@@ -168,9 +210,10 @@ func main() { //nolint:funlen,cyclop
 	}
 
 	if err := (&cpmsgeneratorcontroller.ControlPlaneMachineSetGeneratorReconciler{
-		Client:    mgr.GetClient(),
-		Scheme:    mgr.GetScheme(),
-		Namespace: managedNamespace,
+		Client:                        mgr.GetClient(),
+		Scheme:                        mgr.GetScheme(),
+		Namespace:                     managedNamespace,
+		VSphereCPMSFeatureGateEnabled: vSphereCPMSFeatureGateEnabled,
```
Comment (JoelSpeed) [inline cmd/control-plane-machine-set-operator/main.go:216]: I would probably put the feature gate accessor as a member of the reconciler struct so that we can keep this around once this feature gate is gone/so it can be re-used by other features

File: manifests/0000_30_control-plane-machine-set-operator_00_controlplanemachineset-default.crd.yaml [hunk 23]
```diff
@@ -6,6 +6,7 @@ metadata:
     api-approved.openshift.io: https://github.com/openshift/api/pull/1112
     exclude.release.openshift.io/internal-openshift-hosted: "true"
     include.release.openshift.io/self-managed-high-availability: "true"
+    release.openshift.io/feature-set: Default
```
Comment (JoelSpeed) [inline manifests/0000_30_control-plane-machine-set-operator_00_controlplanemachineset-default.crd.yaml:9]: We need to make sure the TechPreview and Custom versions of this are included as well

File: pkg/controllers/controlplanemachinesetgenerator/controller.go [hunk 24]
```diff
@@ -248,6 +252,16 @@ func (r *ControlPlaneMachineSetGeneratorReconciler) generateControlPlaneMachineS
 		if err != nil {
 			return nil, fmt.Errorf("unable to generate control plane machine set spec: %w", err)
 		}
+	case configv1.VSpherePlatformType:
+		if !r.VSphereCPMSFeatureGateEnabled {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/controller.go:256]: Should move the feature gate accessor `Enabled` call to here IMO

File: pkg/controllers/controlplanemachinesetgenerator/utils.go [hunk 25]
```diff
@@ -167,11 +167,47 @@ func convertViaJSON(in, out interface{}) error {
...
+		if cpmsFailureDomain.OpenStack == nil || cpmsFailureDomain.Platform == "" {
+			return nil, errNoFailureDomains
+		}
+	default:
+		return nil, fmt.Errorf("%w: %sFailureDomain{}", errUnsupportedPlatform, platformType)
+	}
+
+	if cpmsFailureDomain.Platform == "" {
+		return nil, errNoFailureDomains
+	}
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/utils.go:205]: Can we actually get here? Is there ever a generation on AWS/Azure/GCP without failure domains?

File: pkg/controllers/controlplanemachinesetgenerator/utils.go [hunk 26]
```diff
@@ -167,11 +167,47 @@ func convertViaJSON(in, out interface{}) error {
...
-//
-// TO-DO: remove unparam when there are users of the infrastructure parameter
+// buildPlatformFailureDomains returns a failureDomains set for a specific platform.
 //
-//nolint:cyclop,unparam
+//nolint:cyclop
+func buildPlatformFailureDomains(platformType configv1.PlatformType, failureDomains *failuredomain.Set) (*machinev1.FailureDomains, error) {
+	var cpmsFailureDomain machinev1.FailureDomains
+
+	var err error
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/utils.go:176]: Does the linter complain if you remove this empty line?

File: pkg/controllers/controlplanemachinesetgenerator/utils.go [hunk 27]
```diff
@@ -198,30 +236,13 @@ func buildFailureDomains(logger logr.Logger, machineSets []machinev1beta1.Machin
...
-		}
-
-		if cpmsFailureDomain.OpenStack == nil || cpmsFailureDomain.Platform == "" {
-			return nil, nil //nolint:nilnil
-		}
-	default:
-		return nil, fmt.Errorf("%w: %sFailureDomain{}", errUnsupportedPlatform, machineFailureDomains[0].Type())
+	cpmsFailureDomain, err := buildPlatformFailureDomains(platformType, failureDomains)
+	if err != nil {
+		return nil, err
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/utils.go:241]: Error should be wrapped

File: pkg/controllers/controlplanemachinesetgenerator/utils.go [hunk 28]
```diff
@@ -198,30 +236,13 @@ func buildFailureDomains(logger logr.Logger, machineSets []machinev1beta1.Machin
...
 	}
 
-	if cpmsFailureDomain.Platform == "" {
-		return nil, errNoFailureDomains
+	if cpmsFailureDomain == nil && err == nil {
+		return nil, nil //nolint:nilnil
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachinesetgenerator/utils.go:245]: Why is this ok to be nil nil? Can you add a quick comment to explain it? Can we even get a nil nil? Looking at `buildPlatformFailureDomains` I don't see it as possible to return nil nil?

**Comment by jcpowermac** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.15-e2e-vsphere-ovn-serial

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 10 above)
Comment (JoelSpeed) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:134]: Unresolved conversation thread

File: Makefile (hunk 21 above)
Comment (rvanderp3) [reply Makefile:86]: yes, it does.

File: cmd/control-plane-machine-set-operator/main.go (hunk 22 above)
Comment (rvanderp3) [reply cmd/control-plane-machine-set-operator/main.go:216]: cc: @vr4manta

File: pkg/controllers/controlplanemachinesetgenerator/utils.go (hunk 28 above)
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/utils.go:245]: you are correct, removing that statement

File: pkg/controllers/controlplanemachinesetgenerator/utils.go (hunk 26 above)
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/utils.go:176]: yes, it does with 'declarations should never be cuddled'

File: pkg/controllers/controlplanemachinesetgenerator/utils.go (hunk 25 above)
Comment (rvanderp3) [reply pkg/controllers/controlplanemachinesetgenerator/utils.go:205]: it doesn't appear we can, just removed this.

File: pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go (hunk 10 above)
Comment (rvanderp3) [reply pkg/machineproviders/providers/openshift/machine/v1beta1/providerconfig/vsphere.go:134]: Yes, it supports this today.

**Comment by rvanderp3** [comment]
/hold

placing hold for additional work on comments
